tiledb/sm/array_schema/dimension.cc

```diff
@@ -2114,972 +1610,112 @@ void Dimension::set_compute_mbr_func() {
-    case Datatype::DATETIME_DAY:
-    case Datatype::DATETIME_HR:
-    case Datatype::DATETIME_MIN:
-    case Datatype::DATETIME_SEC:
-    case Datatype::DATETIME_MS:
-    case Datatype::DATETIME_US:
-    case Datatype::DATETIME_NS:
-    case Datatype::DATETIME_PS:
-    case Datatype::DATETIME_FS:
-    case Datatype::DATETIME_AS:
-    case Datatype::TIME_HR:
-    case Datatype::TIME_MIN:
-    case Datatype::TIME_SEC:
-    case Datatype::TIME_MS:
-    case Datatype::TIME_US:
-    case Datatype::TIME_NS:
-    case Datatype::TIME_PS:
-    case Datatype::TIME_FS:
-    case Datatype::TIME_AS:
-      map_from_uint64_func_ = map_from_uint64<int64_t>;
-      break;
-    case Datatype::STRING_ASCII:
-      map_from_uint64_func_ = map_from_uint64<char>;
-      break;
-    default:
-      map_from_uint64_func_ = nullptr;
-      break;
-  }
+  auto g = [&](auto T) {
+    map_from_uint64_func_ = map_from_uint64<decltype(T)>;
+  };
+  execute_callback_with_type(type_, g);
 }
 
 void Dimension::set_smaller_than_func() {
-  switch (type_) {
-    case Datatype::INT32:
-      smaller_than_func_ = smaller_than<int32_t>;
-      break;
-    case Datatype::INT64:
-      smaller_than_func_ = smaller_than<int64_t>;
-      break;
-    case Datatype::INT8:
-      smaller_than_func_ = smaller_than<int8_t>;
-      break;
-    case Datatype::UINT8:
-      smaller_than_func_ = smaller_than<uint8_t>;
-      break;
-    case Datatype::INT16:
-      smaller_than_func_ = smaller_than<int16_t>;
-      break;
-    case Datatype::UINT16:
-      smaller_than_func_ = smaller_than<uint16_t>;
-      break;
-    case Datatype::UINT32:
-      smaller_than_func_ = smaller_than<uint32_t>;
-      break;
-    case Datatype::UINT64:
-      smaller_than_func_ = smaller_than<uint64_t>;
-      break;
-    case Datatype::FLOAT32:
-      smaller_than_func_ = smaller_than<float>;
-      break;
-    case Datatype::FLOAT64:
-      smaller_than_func_ = smaller_than<double>;
-      break;
-    case Datatype::DATETIME_YEAR:
-    case Datatype::DATETIME_MONTH:
-    case Datatype::DATETIME_WEEK:
-    case Datatype::DATETIME_DAY:
-    case Datatype::DATETIME_HR:
-    case Datatype::DATETIME_MIN:
-    case Datatype::DATETIME_SEC:
-    case Datatype::DATETIME_MS:
-    case Datatype::DATETIME_US:
-    case Datatype::DATETIME_NS:
-    case Datatype::DATETIME_PS:
-    case Datatype::DATETIME_FS:
-    case Datatype::DATETIME_AS:
-    case Datatype::TIME_HR:
-    case Datatype::TIME_MIN:
-    case Datatype::TIME_SEC:
-    case Datatype::TIME_MS:
-    case Datatype::TIME_US:
-    case Datatype::TIME_NS:
-    case Datatype::TIME_PS:
-    case Datatype::TIME_FS:
-    case Datatype::TIME_AS:
-      smaller_than_func_ = smaller_than<int64_t>;
-      break;
-    case Datatype::STRING_ASCII:
-      smaller_than_func_ = smaller_than<char>;
-      break;
-    default:
-      smaller_than_func_ = nullptr;
-      break;
-  }
+  auto g = [&](auto T) { smaller_than_func_ = smaller_than<decltype(T)>; };
```

Comment:
I still think cases like these should be guarded by a concept that says everything that is supported. That way, we can extent execute_callback_with_type to BLOB and other types not in this PR.